tool/tsh/common/scan.go

```diff
@@ -0,0 +1,246 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package common
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"runtime"
+	"strings"
+
+	"github.com/alecthomas/kingpin/v2"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/constants"
+	"github.com/gravitational/teleport/api/types/accessgraph"
+	secretsscannerclient "github.com/gravitational/teleport/lib/secretsscanner/client"
+	"github.com/gravitational/teleport/lib/secretsscanner/scan"
+
+	accessgraphsecretsv1pb "github.com/gravitational/teleport/api/gen/proto/go/teleport/accessgraph/v1"
+	devicepb "github.com/gravitational/teleport/api/gen/proto/go/teleport/devicetrust/v1"
+	"github.com/gravitational/teleport/lib/devicetrust/assert"
+	dtnative "github.com/gravitational/teleport/lib/devicetrust/native"
+)
+
+type scanCommand struct {
+	keys *scanKeysCommand
+}
+
+func newScanCommand(app *kingpin.Application) scanCommand {
+	scan := app.Command("scan", "Scan the local machine for Secrets and report findings to Teleport")
+	cmd := scanCommand{
+		keys: newScanKeysCommand(scan),
+	}
+	return cmd
+}
+
+type scanKeysCommand struct {
+	*kingpin.CmdClause
+	dirs []string
+	ca   string
+	out  io.Writer
+}
+
+func newScanKeysCommand(parent *kingpin.CmdClause) *scanKeysCommand {
+	c := &scanKeysCommand{CmdClause: parent.Command("keys", "Scan the local machine for SSH private keys and report findings to Teleport")}
+	c.Flag("dirs", "Directories to scan.").Default(defaultDirValues()).StringsVar(&c.dirs)
+	return c
+}
+
+func defaultDirValues() string {
+	switch runtime.GOOS {
+	case constants.LinuxOS:
+		return "/home/"
+	case constants.DarwinOS:
+		return "/Users/"
+	case constants.WindowsOS:
+		return "C:\\Users\\"
+	default:
+		return "/"
+	}
+}
+
+func (c *scanKeysCommand) run(cf *CLIConf) error {
+	ctx := cf.Context
+
+	deviceCred, err := dtnative.GetDeviceCredential()
+	if err != nil {
+		return trace.Wrap(err, "device not enrolled")
+	}
+
+	fmt.Printf("Device trust credentials found.\nScanning %s.\n", strings.Join(c.dirs, ", "))
+
+	scanner, err := scan.NewScanner(scan.ScannerConfig{
+		Dirs: c.dirs,
+		Log:  slog.Default(),
+	})
+	if err != nil {
+		return trace.Wrap(err, "failed to create scanner")
+	}
+
+	privateKeys := scanner.ScanPrivateKeys(
+		ctx,
+		deviceCred.Id,
+	)
+
+	printPrivateKeys(privateKeys)
+
+	client, err := secretsscannerclient.NewSecretsScannerServiceClient(
+		ctx,
+		secretsscannerclient.ClientConfig{
+			ProxyServer: cf.Proxy,
+			Insecure:    cf.InsecureSkipVerify,
+			Log:         slog.Default(),
+		})
+
+	if err := authenticateAndReportPrivateKeys(ctx, client, privateKeys); trace.IsNotImplemented(err) {
```

Comment:
Is it worth moving this to lib/secretsscanner/scan so it can be tested too?

Reply:
moved to reporter